[PyTorch Edge] Support default args with out arg, flag off

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63540

1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

Commit:
[PyTorch Edge] Support default args with out arg



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg"


Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -99,20 +99,23 @@ bool Function::append_operator(
-        for (size_t i = num_specified_args.value(); i < args.size(); ++i) {
-          stack.push_back(args[i].default_value());
+        c10::optional<IValue> out_arg;
+        if (!args.empty() && args.back().is_out()) {
```
Is there always zero or one out arg? What happens if there are 2 or more out args?

cccclai replied:
Oh I didn't know it's possible. My understanding is that out arg is an optional argument to store the output to. The output from inference can only be one. What about having a TORCH_CHECK to check only 1 out arg and we can extend it when we know there might be more than one out arg?

raziel replied:
I think there can be more than 1 out arg. We definitely are coding the out arg as a counter not a bool (plus it makes sense it can be > 1).

cccclai replied:
Update to cover multiple args.

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -1354,6 +1354,82 @@ TEST(LiteInterpreterTest, DefaultArgsPinvSpecifyDefault) {
+    )");
+  }
+
+  // std::vector<torch::jit::IValue> inputs;
```
Just so we don't forget: there's a few debug lines commented out below to remove before landing.

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -1354,6 +1354,82 @@ TEST(LiteInterpreterTest, DefaultArgsPinvSpecifyDefault) {
+  }
+
+  for (auto& it : module_debug_info_set) {
+    std::cout << "one line: " << it << std::endl;
```
I'm guessing this is for debugging?

cccclai replied:
Yeah it's for debugging. Sorry for the noise. Let me clean it up

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -1354,6 +1354,82 @@ TEST(LiteInterpreterTest, DefaultArgsPinvSpecifyDefault) {
+        module_debug_info_set.insert(module_info);
+      }
+      ++pc;
+    } catch (const std::exception& e) {
```
why is the exception the while's exit condition?

cccclai replied:
Sorry it's debugging lines. Removed.

raziel commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -978,11 +978,13 @@ MobileCode::MobileCode(
     const std::shared_ptr<Graph>& graph,
     std::string function_name,
     bool emit_default_input_instructions,
+    bool support_default_args_before_out,
```
why do we need a flag for this? Shouldn't it always be true?

cccclai replied:
It's for backport.

cccclai replied:
Add the backport function in https://github.com/pytorch/pytorch/pull/63651

raziel reviewed (commented):
Thanks. A few comments, but in particular about why the support in the interpreter is under a flag?

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"






Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"






Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"






Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -99,20 +99,26 @@ bool Function::append_operator(
+          out_args.push_back(stack.back());
+          stack.pop_back();
+        }
+        size_t start_index = num_specified_args.value() - out_args.size();
```
@iseeyuan @raziel @cccclai @gmagogsfm I had to do this calculation many places, so I modified CalculateNecessaryArgs utility method to return two things (`pair<int, int>`) where the first one is number of default args and the second one is number of out args. Do you think I should communicate this number of args to mobile by either of following ways:
1. change `num_specified_args: int` to `num_specified_args: pair<int, int>`
2. add another map that says `op_to_num_out_args`.

tugsbayasgalan replied:
yeahh I guess you still have to loop from the end to pop out args. Yeah, I confirm mobile doesn't need num_out_args.

iseeyuan replied:
@tugsbayasgalan Thanks for the confirmation! At least we don't need to change the bytecode schema :D

tugsbayasgalan commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -746,11 +748,12 @@ struct MobileCodeImpl : CodeImpl {
+              support_default_args_before_out_);
+
+          size_t numInclude = specifiedArgs.first +
+              (support_default_args_before_out_ ? specifiedArgs.second : 0);
```
nice!

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -99,20 +99,26 @@ bool Function::append_operator(
+        size_t start_index = num_specified_args.value() - out_args.size();
+        for (size_t i = (start_index >= 0 ? start_index : 0); i < args.size();
+             ++i) {
+          if (args[i].default_value().has_value()) {
```
It should be a TORCH_CHECK. `args[i]` must have default value, otherwise the element# in the stack can mismatch.

cccclai replied:
Resolve in the latest commit.

iseeyuan reviewed (changes requested):
Left a few comments. @cccclai For checking outputs in tests and TORCH_CHECK default_value, please let me know if they make sense.

@tugsbayasgalan Please verify that we don't need explicit `num_out_args`.

cccclai commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -1354,6 +1354,67 @@ TEST(LiteInterpreterTest, DefaultArgsPinvSpecifyDefault) {
+  input[0] = 1; // a more stable matrix
+  input = input.view({N, N});
+  auto ref = m.run_method("forward", input);
+  TORCH_CHECK(!input.equal(torch::range(1, N * N, 1)));
```
Since it is using out arg, it modifies `input` in place and output should be `None`

iseeyuan replied:
Ah, it's in-place calculation. It's interesting that the pseudo inverse of a matrix would be the same as the original matrix. My experiment shows something different: 
```
import torch
N = 3
input = torch.range(1, N * N, 1)
input = input.view(N, N)
print(input)
torch.linalg.pinv(input, 1e-5, out=input)
print(input)
```
```
tensor([[1., 2., 3.],
        [4., 5., 6.],
        [7., 8., 9.]])
tensor([[-6.3889e-01, -1.6667e-01,  3.0556e-01],
        [-5.5555e-02,  1.3039e-08,  5.5556e-02],
        [ 5.2778e-01,  1.6667e-01, -1.9444e-01]])
```

```
tensor([[-6.3889e-01, -1.6667e-01,  3.0556e-01],
        [-5.5555e-02,  1.3039e-08,  5.5556e-02],
        [ 5.2778e-01,  1.6667e-01, -1.9444e-01]])
```

iseeyuan replied:
Oh, NOT equal to is checked. It's checking that "the input is changed in this method". It might be fine for this test.

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

iseeyuan commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -1354,6 +1354,68 @@ TEST(LiteInterpreterTest, DefaultArgsPinvSpecifyDefault) {
+
+  const int N = 28;
+  auto input = torch::range(1, N * N, 1);
+  input[0] = 1; // a more stable matrix
```
input[0] is already 1. This line does not change input. It does not make the condition better. It's still a very ill-conditioned matrix. Maybe change input[0] = 10000 or something.

cccclai replied:
Hmm can you elaborate it a bit more? Do you change `input[0]` to some other number?

iseeyuan replied:
```
import torch
N = 3
input = torch.range(1, N * N, 1)
print(input)
```
```
tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.])
```
`input[0] = 1` does not change input and the condition number of this matrix is Inf (very unstable) and the solution could vary in different platforms.

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -99,21 +99,34 @@ bool Function::append_operator(
+          stack.pop_back();
+        }
+        size_t start_index = num_specified_args.value() - out_args.size();
+        for (size_t i = (start_index >= 0 ? start_index : 0); i < args.size();
```
@raziel , I think your point is to TORCH_CHECK(start_index >= 0).

cccclai replied:
Add the check for `start_index>=0` in latest commit.

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -99,21 +99,34 @@ bool Function::append_operator(
+        for (size_t i = (start_index >= 0 ? start_index : 0); i < args.size();
+             ++i) {
+          auto default_val = args[i].default_value();
+          TORCH_CHECK(
```
@raziel I'm trying to get your point. Do you mean the schema of the model or the schema of the runtime operator? If we also serialize the schema of the operator that make sense but would it increase the model size?

iseeyuan replied:
@raziel Got it. Thanks!

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -99,21 +99,34 @@ bool Function::append_operator(
+
           stack.push_back(args[i].default_value());
         }
+        while (!out_args.empty()) {
```
Good point! Update it in latest commit.

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/runtime/interpreter/code_impl.h:
```diff
@@ -782,21 +785,25 @@ struct MobileCodeImpl : CodeImpl {
   }
 
   bool emit_default_input_instructions_;
+  bool support_default_args_before_out_;
```
We could clean them up when we don't need the backport functions.

iseeyuan reviewed (approved):
LGTM. Thanks!

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

raziel reviewed (approved):
Thanks!

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Support default args with out arg, flag off"


1. Allow consuming operators with defaults arguments and out arguments. Flag is off to keep the same behavior as v6, in pr 63651, turn on the flag.
2. Add two unittests to cover this type of operators.



Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156/)

Differential Revision: [D30414156](https://our.internmc.facebook.com/intern/diff/D30414156)

[ghstack-poisoned]
